File: examples_py/motion/movej.py
#!/usr/bin/env python3
"""motion/movej - joint-space motion with MoveJ.

Usage:
    python examples_py/motion/movej.py [robot_ip]

Safety note:
    This example moves to a fixed conservative joint target. Verify the target
    is safe for your robot before running.
"""

import sys
import time

from rcore_sdk import JointPositions, Robot


def check(result, label):
    if not result:
        raise RuntimeError(
            f"{label} failed: code={result.error_code} msg={result.error_msg}"
        )


def main():
    robot_ip = sys.argv[1] if len(sys.argv) > 1 else ""

    robot = Robot()
    if not robot.Initialize(robot_ip):
        print("Initialize failed", file=sys.stderr)
        return 1

    try:
        check(robot.Enable(), "Enable")
        time.sleep(0.5)

        # Joint velocity percentage is a 6-element list (10% here).
        check(robot.SetVelocityPercentage([0.1] * 6), "SetVelocityPercentage")
        print("Velocity percentage set to 10%")

        # Fixed conservative target [rad]. Check before running on real hardware.
        target = JointPositions([0.0, -1.5708, 1.5708, -1.5708, -1.5708, 0.0])
        print(f"MoveJ target [rad]: {' '.join(f'{v:.4f}' for v in target)}")

        # Synchronous mode: block until the motion finishes. The optional
        # velocity_scale (0,1] multiplies the controller's global velocity for
        # this single motion only; 0.5 here halves it. Omit it to use the
        # global velocity unchanged.
        check(robot.MoveJ(target, velocity_scale=0.5), "MoveJ")
        print("MoveJ completed")
    finally:
        robot.Shutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/motion/movej.cpp
// motion/movej - joint-space motion with MoveJ
//
// Usage: ./motion_movej [robot_ip]
//
// Safety note: this example moves to a fixed conservative joint target.
// Verify this target is safe for your robot before running.

#include "sdk/robot.hpp"

#include <array>
#include <chrono>
#include <cstdio>
#include <exception>
#include <string>
#include <thread>

int main(int argc, char **argv)
{
    const std::string robot_ip = (argc > 1) ? argv[1] : "";

    try
    {
        rcore::sdk::Robot robot;
        if (!robot.Initialize(robot_ip))
        {
            std::fprintf(stderr, "Initialize failed\n");
            return 1;
        }

        // Enable motors and give the controller time to switch modes.
        auto enable_result = robot.Enable();
        if (!enable_result.IsSuccess())
        {
            std::fprintf(stderr, "Enable failed: code=%u msg=%s\n",
                         enable_result.GetErrorCode(),
                         enable_result.GetErrorMsg().c_str());
            robot.Shutdown();
            return 1;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(500));

        const std::array<double, 6> safe_velocity{0.1, 0.1, 0.1,
                                                  0.1, 0.1, 1};
        auto velocity_result =
            robot.Config().SetVelocityPercentage(safe_velocity);
        if (!velocity_result.IsSuccess())
        {
            std::fprintf(stderr,
                         "SetVelocityPercentage(10%%) failed: code=%u msg=%s\n",
                         velocity_result.GetErrorCode(),
                         velocity_result.GetErrorMsg().c_str());
            robot.Shutdown();
            return 1;
        }
        std::printf("Velocity percentage set to 10%%\n");

        // Fixed conservative target [rad]. Check and modify this target before
        // running on a real robot.
        rcore::sdk::JointPositions target{0.0, -1.5708, 1.5708,
                                          -1.5708, -1.5708, 0.0};

        std::printf("MoveJ target [rad]:");
        for (size_t i = 0; i < 6; ++i)
        {
            std::printf(" %.4f", target[i]);
        }
        std::printf("\n");

        // Synchronous mode: block until the motion finishes. The optional
        // velocity_scale (0,1] multiplies the controller's global velocity for
        // this single motion only; 0.5 here halves it. Omit it (or pass -1) to
        // use the global velocity unchanged.
        auto result = robot.MoveJ(target, /*async=*/false,
                                  /*velocity_scale=*/0.5);
        if (!result.IsSuccess())
        {
            std::fprintf(stderr, "MoveJ failed: code=%u msg=%s\n",
                         result.GetErrorCode(), result.GetErrorMsg().c_str());
            robot.Shutdown();
            return 1;
        }
        std::printf("MoveJ completed\n");

        robot.Shutdown();
        return 0;
    }
    catch (const std::exception &e)
    {
        std::fprintf(stderr, "Exception: %s\n", e.what());
        return 1;
    }
}
